internal/pubsub/pubsub.go

```diff
@@ -0,0 +1,138 @@
+package pubsub
+
+import (
+	"context"
+	"encoding/json"
+	"net/url"
+	"regexp"
+
+	cloudpubsub "cloud.google.com/go/pubsub"
+	"github.com/GoogleCloudPlatform/cloud-run-release-operator/internal/health"
+	"github.com/GoogleCloudPlatform/cloud-run-release-operator/internal/util"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/api/run/v1"
+)
+
+// Client represents a client to Google Cloud Pub/Sub.
+type Client interface {
+	Publish(ctx context.Context, msg Message) error
+}
+
+// PubSub is a Google Cloud Pub/Sub client to publish messages.
+type PubSub struct {
+	client *cloudpubsub.Client
+	topic  *cloudpubsub.Topic
+}
+
+// Message is the format when publishing to Pub/Sub.
+type Message struct {
+	Event                        string       `json:"event"`
+	CandidateRevisionName        string       `json:"candidateRevisionName"`
+	CandidateRevisionPercent     int          `json:"candidateRevisionPercent"`
+	CandidateRevisionURL         string       `json:"candidateRevisionURL"`
+	CandidateWasPromotedToStable bool         `json:"candidateWasPromotedToStable"`
+	Service                      *run.Service `json:"service"`
+}
+
+// New initializes a PubSub client.
+func New(ctx context.Context, projectID string, topicName string) (ps PubSub, err error) {
+	logger := util.LoggerFrom(ctx)
+	client, err := cloudpubsub.NewClient(ctx, projectID)
+	if err != nil {
+		return ps, errors.Wrap(err, "failed to initialize Pub/Sub client")
+	}
+
+	match := regexp.MustCompile(`projects/([^/]*)/topics/([^/]*)`).FindStringSubmatch(topicName)
+	if len(match) != 3 {
+		return ps, errors.Errorf("invalid topic name %s", topicName)
+	}
+	project := match[1]
+	topicID := match[2]
+	logger.WithFields(logrus.Fields{"topicProject": project, "topicID": topicID}).Debug("parsed topic name value")
+
+	return PubSub{
+		client: client,
+		topic:  client.TopicInProject(topicID, project),
+	}, nil
+}
+
+// NewMessage initializes a message.
+func NewMessage(svc *run.Service, diagnosis health.DiagnosisResult, candidateWasPromoted bool) (Message, error) {
+	event := "rollout"
+	if diagnosis == health.Unhealthy {
+		event = "rollback"
+	}
+
+	var candidateRevision *run.TrafficTarget
+	var err error
+	if candidateWasPromoted {
+		// The last candidate is now the stable.
+		candidateRevision, err = findRevisionWithTag(svc, "stable")
+	} else {
+		candidateRevision, err = findRevisionWithTag(svc, "candidate")
+	}
```

Comment:
it's a bit weird we re-evaluate the type and compute things again here.

Reply:
don't we have these already computed in rollout code?

Reply:
We do, but we only get the revision name. There's also the problem that the info about the revision is obtained before rolling out/back, so it became outdated by the time things are published to PubSub.

That's without mentioning the problem outlined in the comment for `findRevisionWithTag`